docs/src/proposals/stake-v2.md

```diff
@@ -0,0 +1,215 @@
+    redelegation_epoch: Epoch,
+    credits_observed: u64,
+}
+```
+
+With this information, we can calculate multi-epoch redelegations when the cluster
+sees more than 25% stake movement.
+
+### Interoperability with V1
+
+Since loads of places use `mem::size_of::<StakeState>()`, and we still need
+to support that, that struct won't get touched. On the flip side, programs
+need to process both. Since there's already a neat interface on `StakeState`,
+we can do something similar with a new type:
+
+```
+enum StakeStateV2 {
+    Uninitialized,
+    Initialized(Meta),
+    StakeV1(Meta, Stake),
+    RewardsPool,
+    StakeV2(StakeV2),
+}
+impl StakeStateV2 {
+    /// ... all the same as StakeState ...
+    pub fn redelegation(&self) -> Option<Redelegation> {
+      match self {
+          Self::StakeV2(stake) => Some(stake.redelegation),
+          _ => None,
+      }
+    }
+}
+```
+
+Programs update to use `StakeStateV2` everywhere, and existing code still works.
+When an `Initialized` stake is delegated, the stake program makes it a `StakeV1`
+if the size is `200`, and a `StakeV2` if the size is `256`.
+
+### Epoch Boundary Calculations
+
+While we're at it, new stake accounts can also solve point number 2, by having
+their rewards paid through an update instruction, and not automatically at every
+epoch boundary.
+
+During epoch boundary, the validator vote accounts receive all stake V2
+account rewards as withheld lamports in their account. They cannot
+touch these, and stake accounts may claim them.
+
+The biggest users of stake accounts, ie stake pools and stake bots, already
+touch stake accounts every epoch during their bookkeeping, so this adds trivial
+work for them.
+
+Legacy stake accounts are unaffected, and inflation continues to pay out their rewards
+at epoch boundaries automatically.
+
+## New Stake Program Instructions
+
+`Split` / `Merge` / `Deactivate` are only allowed if the stake account has
+been updated this epoch, or after deactivation epoch if applicable.
+
+Side note: deactivate might actually be OK, since we could technically calculate
+the rewards even after deactivation.
+
+### UpgradeToV2
+
+Given a V1 stake account, a signature from the stake or withdraw authority, and
+a funding account, reallocate the existing stake account and convert it to a V2
+stake. The funding account adds the lamports necessary to cover the new rent-exemption.
+
+Side note: this would also be a great time to include the rent-exemption in the
+delegation. That way, smaller stakers get a tiny bit more, and there's a couple of other
+nice side effects. Ask me about it.
+
+### Redelegate
+
+Given a V2 stake account, vote account, and stake authority signature, change the
+redelegation. Follows most of the same rules as delegate.
+
+If it's going through a multi-epoch redelegation, be sure that previous epoch
+rewards have been claimed, and update the vote pubkey / credits / epoch.
+
+### (Also in the Vote Program) ClaimRewards
```

Comment:
I'm a big fan of this direction. Giving users the tools to decide when rewards are claimed can be very valuable from a tax perspective (speaking from a US understanding). Additionally, lazy computation of rewards enables spreading out the calculation over any slot in an epoch, which you've identified in issue 2.